### PR TITLE
Fixed #25307 -- conditional aggregation over subqueries

### PR DESCRIPTION
**Pull request by akaariai**

### CODE REVIEW

File: tests/aggregation_regress/tests.py [hunk 1]
```diff
@@ -360,6 +361,75 @@ def test_aggrate_annotation(self):
...
+        # grouped queryset.
+        self.assertEqual(
+            Book.objects.annotate(c=Count('authors')).values('c').aggregate(test=Sum(
+                Case(When(c__gt=1, then=1), output_field=IntegerField())
+            ))['test'],
+            3
+        )
+
+    def test_sliced_conditional_aggregate(self):
+        """
```
Comment (timgraham) [inline tests/aggregation_regress/tests.py]: I think these docstrings aren't needed.

File: tests/aggregation_regress/tests.py [hunk 2]
```diff
@@ -360,6 +361,75 @@ def test_aggrate_annotation(self):
             {'c__max': 3}
         )
 
+        # Regression for #25307:  AttributeError: 'WhereNode' object has no
+        # attribute 'get_source_expressions'. Conditional aggregation of a
+        # grouped queryset.
+        self.assertEqual(
+            Book.objects.annotate(c=Count('authors')).values('c').aggregate(test=Sum(
```
Comment (timgraham) [inline tests/aggregation_regress/tests.py]: This isn't passing on any backends. SQLite, for example: `OperationalError: misuse of aggregate function COUNT()`.

**Comment by timgraham** [comment]
It's fine to squash commits and write "Thanks Travis Newport for the tests." in the commit message.

**Comment by newportt** [comment]
I believe the `Lookup`s will have to be added as annotations to the subquery. The fix would be to change [this line](https://github.com/django/django/blob/master/django/db/models/sql/query.py#L361) to:`elif isinstance(expr, (Col, Lookup)):`

**Comment by timgraham** [comment]
There is still trouble in `aggregation_regress` on Oracle before and after that suggested change (before `ORA-00979: not a GROUP BY expression`, after `ORA-00920: invalid relational operator`).

**Comment by timgraham** [comment]
Oracle failures (only checked `aggregation_regress` locally):
```
======================================================================
ERROR: test_aggregate_annotation (aggregation_regress.tests.AggregationTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/oracle/base.py", line 481, in execute
    return self.cursor.execute(query, self._param_generator(params))
cx_Oracle.DatabaseError: ORA-00920: invalid relational operator


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tim/code/django/tests/aggregation_regress/tests.py", line 368, in test_aggregate_annotation
    Case(When(c__gt=1, then=1), output_field=IntegerField())
  File "/home/tim/code/django/django/db/models/query.py", line 359, in aggregate
    return query.get_aggregation(self.db, kwargs.keys())
  File "/home/tim/code/django/django/db/models/sql/query.py", line 457, in get_aggregation
    result = compiler.execute_sql(SINGLE)
  File "/home/tim/code/django/django/db/models/sql/compiler.py", line 835, in execute_sql
    cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/tim/code/django/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/tim/code/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/oracle/base.py", line 481, in execute
    return self.cursor.execute(query, self._param_generator(params))
django.db.utils.DatabaseError: ORA-00920: invalid relational operator


======================================================================
ERROR: test_annotated_conditional_aggregate (aggregation_regress.tests.AggregationTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/oracle/base.py", line 481, in execute
    return self.cursor.execute(query, self._param_generator(params))
cx_Oracle.DatabaseError: ORA-00979: not a GROUP BY expression


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tim/code/django/tests/aggregation_regress/tests.py", line 386, in test_annotated_conditional_aggregate
    output_field=DecimalField()
  File "/home/tim/code/django/django/db/models/query.py", line 359, in aggregate
    return query.get_aggregation(self.db, kwargs.keys())
  File "/home/tim/code/django/django/db/models/sql/query.py", line 457, in get_aggregation
    result = compiler.execute_sql(SINGLE)
  File "/home/tim/code/django/django/db/models/sql/compiler.py", line 835, in execute_sql
    cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/tim/code/django/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/tim/code/django/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/tim/code/django/django/db/backends/oracle/base.py", line 481, in execute
    return self.cursor.execute(query, self._param_generator(params))
django.db.utils.DatabaseError: ORA-00979: not a GROUP BY expression
```

**Comment by timgraham** [comment]
buildbot, test on oracle.

File: django/db/models/sql/where.py [hunk 3]
```diff
@@ -160,6 +170,13 @@ def _contains_aggregate(cls, obj):
     def contains_aggregate(self):
         return self._contains_aggregate(self)
 
+    @property
+    def is_summary(self):
+        for obj in self.children:
+            if obj.is_summary:
+                return True
+        return False
```
Comment (charettes) [inline django/db/models/sql/where.py]: I find `return any(child.is_summary for child in self.children)` more readable.

**Comment by timgraham** [comment]
@shaib @jarshwah -- can you spot what's making Oracle unhappy?

For the "ORA-00920: invalid relational operator":
Oracle:
```sql
SELECT SUM(CASE WHEN "__COL1" THEN :arg0 ELSE NULL END) FROM (
SELECT "AGGREGATION_REGRESS_BOOK"."ID" AS Col1, 
COUNT("AGGREGATION_REGRESS_BOOK_AC81D"."AUTHOR_ID") > :arg1 AS "__COL1" FROM
 "AGGREGATION_REGRESS_BOOK" LEFT OUTER JOIN 
"AGGREGATION_REGRESS_BOOK_AC81D" ON ("AGGREGATION_REGRESS_BOOK"."ID" =
 "AGGREGATION_REGRESS_BOOK_AC81D"."BOOK_ID") GROUP BY 
"AGGREGATION_REGRESS_BOOK"."ID", "AGGREGATION_REGRESS_BOOK"."ISBN", 
"AGGREGATION_REGRESS_BOOK"."NAME", "AGGREGATION_REGRESS_BOOK"."PAGES",
"AGGREGATION_REGRESS_BOOK"."RATING", "AGGREGATION_REGRESS_BOOK"."PRICE", 
"AGGREGATION_REGRESS_BOOK"."CONTACT_ID", 
"AGGREGATION_REGRESS_BOOK"."PUBLISHER_ID", 
"AGGREGATION_REGRESS_BOOK"."PUBDATE") subquery'
- PARAMS = (1, 1); args=(1, 1)
```
PostgreSQL
```sql
SELECT SUM(CASE WHEN "__col1" THEN 1 ELSE NULL END) FROM (
SELECT "aggregation_regress_book"."id" AS Col1,
 COUNT("aggregation_regress_book_authors"."author_id") > 1 AS "__col1" FROM
 "aggregation_regress_book" LEFT OUTER JOIN "aggregation_regress_book_authors" ON
 ("aggregation_regress_book"."id" = "aggregation_regress_book_authors"."book_id") GROUP BY
 "aggregation_regress_book"."id") subquery;
args=(1, 1)
```
For the "ORA-00979: not a GROUP BY expression":
Oracle:
```sql
SELECT AVG(CASE WHEN "__COL1" < :arg0 THEN "DISCOUNT_PRICE" ELSE NULL END)
FROM (SELECT "AGGREGATION_REGRESS_BOOK"."ID" AS Col1,
("AGGREGATION_REGRESS_BOOK"."PRICE" * :arg1) AS "DISCOUNT_PRICE",
"AGGREGATION_REGRESS_BOOK"."PAGES" AS "__COL1"
FROM "AGGREGATION_REGRESS_BOOK"
GROUP BY "AGGREGATION_REGRESS_BOOK"."ID",
("AGGREGATION_REGRESS_BOOK"."PRICE" * :arg2), "AGGREGATION_REGRESS_BOOK"."PAGES") subquery'
- PARAMS = (400, 0.75, 0.75); args=(400, 0.75, 0.75)
```
PostgreSQL:
```sql
SELECT AVG(CASE WHEN "__col1" < 400 THEN "discount_price" ELSE NULL END)
FROM (SELECT "aggregation_regress_book"."id" AS Col1,
("aggregation_regress_book"."price" * 0.75) AS "discount_price",
"aggregation_regress_book"."pages" AS "__col1" FROM "aggregation_regress_book"
GROUP BY "aggregation_regress_book"."id", ("aggregation_regress_book"."price" * 0.75)) subquery;
args=(400, 0.75, 0.75)

**Comment by shaib** [comment]
Oracle 11 for sure does not support Booleans as values (although I think I saw mentions of them being added in Oracle 12) so `COUNT("AGGREGATION_REGRESS_BOOK_AC81D"."AUTHOR_ID") > :arg1 AS "__COL1"` needs to be changed whether or not it is the cause of the ORA-00920. It should be wrapped in `CASE WHEN (comparison) THEN 1 ELSE 0 END`.

For the ORA-00979, my guess is that it's the difference between `"AGGREGATION_REGRESS_BOOK"."PRICE" * :arg1` in the select-list and `"AGGREGATION_REGRESS_BOOK"."PRICE" * :arg2` in the group-by list. If this is indeed the case, the only way around it would be to use named arguments -- and that, I suspect, would require open heart surgery for the ORM. I thought trying to remove the expressions and just keep the columns in the group-by list might work, but then I thought of grouping by a sum of two columns -- grouping by the single columns would return wrong results for that.

**Comment by shaib** [comment]
Yes, the ORA-00920 is caused by an attempt to use Booleans in SQL. Oracle 12 supports Booleans, but only in PL/SQL (procedures etc), not SQL proper: http://docs.oracle.com/database/121/LNPLS/datatypes.htm#LNPLS348. The error is actually triggered by `CASE WHEN "__col1" THEN ...`, which has no comparison.

**Comment by timgraham** [comment]
Would you object to marking these as `expectedFailure`on Oracle and opening a separate ticket for an Oracle expert to resolve them (either by allowing them to work or by raising a helpful message saying that the query is unsupported on Oracle)?

**Comment by timgraham** [comment]
@shaib, the last question is for you.

**Comment by shaib** [comment]
@timgraham I'm ok with that for the ora-00979 (expression containing variable used twice, identified as different expressions) issue, not so much for the ora-00920 (use of Booleans). The reason for disagreeing on Booleans is that Oracle may be the only core database with the issue, but last I've seen, SQL Server (for example) shared that trait, and so I think we should generally be prepared to handle backends which do not support a native Boolean type.

@manfre might have valuable input here.

**Comment by samuelmaudo** [comment]
Is there any possibility of applying this patch to Django 1.10? I'm migrating a project from Django 1.6 but this bug is stopping me. My code works correctly on Django 1.6 but on Django 1.8, 1.9 and 1.10b1 it raises an AttributeError: ``'WhereNode'` object has no attribute 'get_source_expressions'``.

**Comment by timgraham** [comment]
@samuelmaudo It would be nice to see your test case given that all the tests here are using `Case`/`When` which is new in Django 1.9. Perhaps we can solve your issue with a less invasive patch that doesn't run into the Oracle issues blocking this patch.

**Comment by samuelmaudo** [comment]
This is the test: [complex_aggregate.zip](https://github.com/django/django/files/370133/complex_aggregate.zip)

Curiously, the error is only raised when the ``distinct`` method is used. So I have created a test with it and another without it.

**Comment by timgraham** [comment]
I'm not seeing how this can be considered a regression from 1.6 given the use of the `Case` / `When` functions which don't exist there. Can you provide the 1.6 version of the test?

**Comment by samuelmaudo** [comment]
I used ``django-aggregate-if`` in Django 1.6. I can give you the test for Django 1.6 with that dependency.

**Comment by timgraham** [comment]
Oh, I guess it's probably not possible to develop some a more minimal fix then (feel free to offer one if you find one). Even if we had a working patch, I don't think it qualifies for a backport to 1.10 at this point in the release cycle.

**Comment by samuelmaudo** [comment]
If I have understood well, this patch works fine except with the Oracle backend. So surely I will apply it to my copy of Django. I would like to help but I have never worked with Oracle.

**Comment by jproffitt** [comment]
I would like to help with this if I can. But I know dealing with Oracle is a bit of a pain. Do you have any suggestions for setting up and running tests on Oracle? Or do I need to leave this for an Oracle expert?

**Comment by charettes** [comment]
@jproffitt, https://code.djangoproject.com/wiki/OracleTestSetup worked great for me.

**Comment by timgraham** [comment]
@shaib, what's the SQL supposed to look like for the `SELECT SUM(CASE WHEN "__COL1" THEN :arg0 ELSE NULL END) FROM ... ` case?

**Comment by shaib** [comment]
@timgraham when there's no native boolean support, `SELECT SUM(CASE WHEN "__COL1"=1 THEN :arg0 ELSE NULL END) FROM ...` (note the `=1` added). This assumes Booleans are emulated by ints, which is the normal case; but for this to work here, we'd also need to fix the expression selected as `__COL1`. Quoting myself from above: 
>  `COUNT("AGGREGATION_REGRESS_BOOK_AC81D"."AUTHOR_ID") > :arg1 AS "__COL1"` needs to be [...] wrapped in `CASE WHEN (comparison) THEN 1 ELSE 0 END`

**Comment by timgraham** [comment]
Josh or Anssi, I wonder if you could provide any guidance about where the fix to modify the SQL/expressions should be added (without too much effort)? I'm slowly working my way through the ORM to try to understand how the query is constructed but I thought you mind be able to save me some time with a pointer.

**Comment by jarshwah** [comment]
Tim, are you talking about the boolean expression for Oracle? I haven't got a whole lot of time at the moment unfortunately, but if you need to adjust the SQL fragment in When clauses for boolean expressions, I would probably look at implementing `as_oracle` for `When()` or creating a separate `exact` lookup for boolean expressions that implements `as_oracle`.

Something like:

```
def as_oracle(self, ..):
    if self.output_field.get_field_type() == 'BooleanField':
        return '%s=1', param
```

That's from the top of my head, but that's where I'd begin looking to fix. If I've misunderstood let me know and I'll see what else I can come up with.

The GROUP BY expressions seem weird to me. It's using the wrong parameters by the look of things. That'd take quite a bit more to fix I think.

**Comment by timgraham** [comment]
The original (broken) query is:
```
SELECT SUM(CASE WHEN "__COL1" THEN :arg0 ELSE NULL END) FROM (
SELECT "AGGREGATION_REGRESS_BOOK"."ID" AS Col1, 
COUNT("AGGREGATION_REGRESS_BOOK_AC81D"."AUTHOR_ID") > :arg1 AS "__COL1" FROM
 "AGGREGATION_REGRESS_BOOK" LEFT OUTER JOIN 
"AGGREGATION_REGRESS_BOOK_AC81D" ON ("AGGREGATION_REGRESS_BOOK"."ID" =
 "AGGREGATION_REGRESS_BOOK_AC81D"."BOOK_ID") GROUP BY 
"AGGREGATION_REGRESS_BOOK"."ID", "AGGREGATION_REGRESS_BOOK"."ISBN", 
"AGGREGATION_REGRESS_BOOK"."NAME", "AGGREGATION_REGRESS_BOOK"."PAGES",
"AGGREGATION_REGRESS_BOOK"."RATING", "AGGREGATION_REGRESS_BOOK"."PRICE", 
"AGGREGATION_REGRESS_BOOK"."CONTACT_ID", 
"AGGREGATION_REGRESS_BOOK"."PUBLISHER_ID", 
"AGGREGATION_REGRESS_BOOK"."PUBDATE") subquery'- PARAMS = (1, 1); args=(1, 1)
```
Shai suggested these changes:
```
SELECT SUM(CASE WHEN "__COL1"=1 THEN :arg0 ELSE NULL END) FROM (
SELECT "AGGREGATION_REGRESS_BOOK"."ID" AS Col1, 
CASE WHEN 
   (COUNT("AGGREGATION_REGRESS_BOOK_AC81D"."AUTHOR_ID") > :arg1)
   THEN 1 ELSE 0 
AS "__COL1" FROM
```
(adding `=1` on line 1 and wrapping `CASE WHEN (...) THEN 1 ELSE 0` around `COUNT("AGGREGATION_REGRESS_BOOK_AC81D"."AUTHOR_ID") > :arg1`.

Looking at adding `=1` first, the SQL is built like this:
```
              "__COL1"                                  Ref.as_sql()
         WHEN "__col1" THEN %s, [1]                     When.as_sql()
    CASE WHEN "__col1" THEN %s ELSE NULL END, [1]       Case.as_sql()
SUM(CASE WHEN "__COL1" THEN :arg0 ELSE NULL END)        Func.as_sql()
```
I'm not sure where to go from there. The `When` doesn't have an `output_field`.

**Comment by browniebroke** [comment]
Not sure if that's meant to be supported but I found a corner case which I think gives incorrect results. 
- I've applied this patch against stable/1.8.x
- Tests seem to pass
- BUT when used in my project, the numbers in the annotations are not the ones I expect.

Maybe my query is wrong, maybe this is because I'm still on 1.8, but I've managed to create a tests case to reproduce my problem on 1.8, but I cannot make it run on my machine against the master branch. I thought you might be interested in adding it to this PR to cover it if supported:

```python
    def test_annotated_m2m_conditional_aggregate(self):
        annotated_qs = Book.objects.annotate(pages_per_author=F('pages') / Count('authors'))
        aggregations = annotated_qs.aggregate(
            author8=Sum(Case(
                When(authors__id=8, then='pages_per_author'),
                output_field=IntegerField(), default=0
            ))
        )
        self.assertEqual(
            aggregations['author8'],
            1512 # = 1132/2 + 946. When run, it equals 2078 (= 1132 + 946)
        )
```

I've verified that in `annotated_qs`, the books have the correct value for `pages_per_author`. I'm happy to open a separate issue if this is out of scope.

**Comment by jproffitt** [comment]
@timgraham I'm not sure, but in `When.as_sql`, there is a `self.condition` variable. I believe `self.condition` is a `Node` (?) which might have an `output_field`. 

I'm completely guessing here, but could we potentially check `self.condition.output_field == BooleanField`? I believe `condition_sql` in the line `condition_sql, condition_params = compiler.compile(self.condition)` would be `"__COL1"` in this case.

Of course, we might have to add the `=1` further down in the `compiler.compile`. I wish I knew more about this. Maybe I can get Oracle setup sometime and run some tests.

**Comment by timgraham** [comment]
@jproffitt, `self.condition.output_field` gives `AttributeError: 'WhereNode' object has no attribute '_output_field_or_none'`. See [OracleTestSetup](https://code.djangoproject.com/wiki/OracleTestSetup) for a way to run the tests on Oracle.

**Comment by timgraham** [comment]
I spent several hours loking at `sql.Query.get_aggregation()` and tracking how `SUM(CASE WHEN "__COL1" THEN :arg0 ELSE NULL END)` is moved from the inner query to the outer query. I couldn't determine how this problem could/should be solved or even that it's feasible, so I have low motivation to continue investigating.

I also tried constructing a different queryset that might work on Oracle: `Book.objects.annotate(c=Count('authors')).annotate(c2=Case(When(c__gt=1, then=1), default=0, output_field=IntegerField())).values('pk').aggregate(test=Sum(Case(When(c2=1, then=1), default=0, output_field=IntegerField())` but this still doesn't generate correct SQL. 

Given that nothing is broken on Oracle that wasn't broken before, I don't see much disadvantage to merging this as-is so that other databases can use the functionality. There are several weeks until the 1.11 feature freeze, so anyone interested in Oracle is welcome to give it a try before then or raise an objection to merging this as-is. Thanks!

**Comment by jarshwah** [comment]
I agree that it's fine to merge as-is for the reasons you've given. Our policy will allow a patch to fix this for Oracle in bug fix releases correct? If that's the case, I don't see (much of) an issue fixing this for the backends we can fix it for if we haven't got the support required to specifically fix Oracle.

File: django/db/models/lookups.py [hunk 4]
```diff
@@ -55,6 +55,17 @@ def batch_process_rhs(self, compiler, connection, rhs=None):
             sqls, sqls_params = ['%s'] * len(params), params
         return sqls, sqls_params
 
+    def get_source_expressions(self):
+        if self.rhs_is_direct_value():
+            return [self.lhs]
+        return [self.lhs, self.rhs]
```
Comment (jarshwah) [inline django/db/models/lookups.py]: I'd love to see some comments here and the set_source_expressions method below that tells us why this is necessary.

**Comment by timgraham** [comment]
Yes, I think as long as the fix isn't too invasive, we could consider a backport to a bug fix release under the "bug in new feature" criteria.

@felixxm, if you have time and interest, here's a non-trivial Oracle issue.

**Comment by jarshwah** [comment]
Looking at the queries for `invalid relational operator`, it looks to me as if the queries being generated are incorrect, even for postgres. I'm copying the postgres query below, along with the orm expression that generates it.

```
SELECT SUM(CASE WHEN "__col1" THEN 1 ELSE NULL END) FROM (
SELECT "aggregation_regress_book"."id" AS Col1,
 COUNT("aggregation_regress_book_authors"."author_id") > 1 AS "__col1" FROM
 "aggregation_regress_book" LEFT OUTER JOIN "aggregation_regress_book_authors" ON
 ("aggregation_regress_book"."id" = "aggregation_regress_book_authors"."book_id") GROUP BY
 "aggregation_regress_book"."id") subquery;
```

```
Book.objects.annotate(c=Count('authors')).values('pk').aggregate(test=Sum(
    Case(When(c__gt=1, then=1), output_field=IntegerField())
))
```

The query this should product (I think) is this:

```
SELECT SUM(CASE WHEN "__col1" > 1 THEN 1 ELSE NULL END) FROM (
SELECT "aggregation_regress_book"."id" AS Col1,
 COUNT("aggregation_regress_book_authors"."author_id") AS "__col1" FROM
 "aggregation_regress_book" LEFT OUTER JOIN "aggregation_regress_book_authors" ON
 ("aggregation_regress_book"."id" = "aggregation_regress_book_authors"."book_id") GROUP BY
 "aggregation_regress_book"."id") subquery;
```

Note how the condition `> 1` is in the outer SUM aggregate rather than the internal `COUNT` aggregate. Django is probably resolving the condition incorrectly, but I haven't looked deeply into why.

If we can fix the query for postgres, it'll fix the query for Oracle. Unless someone can tell me the query currently being generated is correct, I don't think this patch is ok to merge anymore.

**Comment by jarshwah** [comment]
The "ORA-00979: not a GROUP BY expression is really strange. I'd expect the following to fail in the same way under Oracle if it's the arguments being presented. I don't (currently) have access to Oracle, but if someone is keen to look, they could try.

```
Book.objects.annotate(discount_price=F('price') * 0.75).values('discount_price').annotate(sum_discount=Sum('discount_price'))
```

That *should* force the `0.75` argument into the SELECT and GROUP BY clause, triggering the same bug, if it is indeed a bug in parameter substitution. If that does fail, then I'd try swapping out `0.75` for `2` to avoid potential decimal point ambiguities. I'd also try writing the exact query without args into a `connection.cursor()` directly, and again with args directly into the cursor. Should try to use the cx_Oracle cursor directly, to see if the bug is with cx_Oracle, or with our own cursor/param implementation.

I'll probably come back to this in a few weeks (going away until after new years) if no one else beats me to it.

**Comment by jarshwah** [comment]
@timgraham what method are you using to print the queries and parameters?

**Comment by jarshwah** [comment]
Ok my investigation into the `"ORA-00979: not a GROUP BY expression"` is a quirk (bug) with Oracle itself. It looks like the GROUP BY is evaluated before argument substitution, and since the variable names are different, it bails out. I replicated the problem with cx_Oracle cursor directly, and also using SQL Developer.

Here's a query in SQL Developer that prompts for parameter values:

```
SELECT ("AGGREGATION_REGRESS_BOOK"."PRICE" * :a1) AS "DISCOUNT_PRICE", 
sum(("AGGREGATION_REGRESS_BOOK"."PRICE" * :a2)) AS "SUM_DISCOUNT" 
FROM django_test."AGGREGATION_REGRESS_BOOK" "AGGREGATION_REGRESS_BOOK"
GROUP BY ("AGGREGATION_REGRESS_BOOK"."PRICE" * :a3), 
"AGGREGATION_REGRESS_BOOK"."NAME" 
ORDER BY "AGGREGATION_REGRESS_BOOK"."NAME" ASC;
```

This fails in the same way. If we change `a3` to `a1`, the query compiles and runs fine.

Summary: this patch didn't cause this bug, it's always been there, we just didn't have a test for it before. I'll create a new ticket summarising my findings.

**Comment by timgraham** [comment]
I probably used print statements right before `cursor.execute()`.

**Comment by shaib** [comment]
@jarshwah Oracle statements can always be re-used with new parameters, that's a feature; given that, checking that the statement is correct (doesn't select non-summary fields outside the group by) for all parameter values is quite reasonable.

Lest it look like I am defending Oracle here, the real bug in its behavior -- a huge missing feature -- is sane treatment of functional dependencies. As far as I know, PG is the only one doing anything sensible in this area. With such treatment, there should be no problem even if the actual argument values are different.

**Comment by timgraham** [comment]
Anssi, please reopen this if you return to it.

**Comment by jproffitt** [comment]
@timgraham is the original issue fixed now for all the other db backends? Or are we still waiting on an oracle fix?

**Comment by timgraham** [comment]
As far as I know, no one has investigated Josh's comment about the SQL being incorrect (for all backends).

**Comment by jarshwah** [comment]
For reference, [https://github.com/django/django/pull/6267#issuecomment-269056618](this comment) is the outstanding bug Tim is referring to.

**Comment by jproffitt** [comment]
@jarshwah I looked into what you said. I found that this new method that was added on `Lookup` is the culprit:

    @property
    def is_summary(self):
        return self.lhs.is_summary or getattr(self.rhs, 'is_summary', False)

I admit that I don't know much about this at all, but I wonder if that method is correct? In the test, it returns `False`. If I change it to always return `True`, the correct SQL that you mentioned will be generated. (see below)

    SELECT SUM(CASE WHEN "__col1" > 1 THEN 1 ELSE NULL END) FROM (
    SELECT "aggregation_regress_book"."id" AS Col1,
      COUNT("aggregation_regress_book_authors"."author_id") AS "__col1" FROM
      "aggregation_regress_book" LEFT OUTER JOIN "aggregation_regress_book_authors" ON
      ("aggregation_regress_book"."id" = "aggregation_regress_book_authors"."book_id") GROUP BY
      "aggregation_regress_book"."id", "aggregation_regress_book"."isbn",
      "aggregation_regress_book"."name", "aggregation_regress_book"."pages",
      "aggregation_regress_book"."rating", "aggregation_regress_book"."price",
      "aggregation_regress_book"."contact_id", "aggregation_regress_book"."publisher_id",
      "aggregation_regress_book"."pubdate") subquery

(I don't know why it add all the fields into the `GROUP BY`, but it does that without my change too, so I'm not sure what that's all about. Maybe I'm logging the query incorrectly.)

With the change, (always returning `False`) the entire django test suite still passes. But of course, I don't know if there is a reason it should return `True` sometime.

@akaariai, is there a reason you did `return self.lhs.is_summary or getattr(self.rhs, 'is_summary', False)`?

I'd love to get this fixed before 1.11 if that is still possible. If you want me to create a new PR with the change, let me know.

**Comment by jarshwah** [comment]
@jproffitt good find! Changing the default to `True` is likely incorrect, but fixing the underlying problem of why the `rhs.is_summary` does not exist is probably the right fix. I'll try and have a play now.

With regard to the GROUP BY fields, it's dependent on the test database. Postgres and MySQL are able to reduce the GROUP BY down to the primary key of each table involved. I'm guessing you're testing with sqlite?

**Comment by jarshwah** [comment]
The problem is somewhere here in `django.db.models.sql.Query.rewrite_cols()`

```
orig_exprs = annotation.get_source_expressions()
        new_exprs = []
        for expr in orig_exprs:
            if isinstance(expr, Ref):
                # Its already a Ref to subquery (see resolve_ref() for
                # details)
                new_exprs.append(expr)
            elif isinstance(expr, Col) or (expr.contains_aggregate and not expr.is_summary):
                # Reference to column. Make sure the referenced column
                # is selected.
                col_cnt += 1
                col_alias = '__col%d' % col_cnt
                self.annotations[col_alias] = expr
                self.append_annotation_mask([col_alias])
                new_exprs.append(Ref(col_alias, expr))
            else:
                # Some other expression not referencing database values
                # directly. Its subexpression might contain Cols.
                new_expr, col_cnt = self.rewrite_cols(expr, col_cnt)
                new_exprs.append(new_expr)
        annotation.set_source_expressions(new_exprs)
        return annotation, col_cnt
```

Specifically, the `elif` condition. When you changed the is_summary default to True, it falls through to the `else` clause, and keeps rewriting all the expressions. Otherwise, when is_summary returns `False` as it's doing now, the whole lookup gets pushed into the inner query rather than just the column reference.'

I *think* we need an additional `elif` condition in this block or an additional condition in the existing elif, but I'm not sure what.

@timgraham what are your thoughts on merging this patch for alpha, and then fixing the bug before beta? If that's not acceptable, would this qualify for a bug fix patch? Identifying the source of the problem has given me greater confidence that the bug can be fixed before a final cut.

**Comment by timgraham** [comment]
If something is working by Monday, I'll look at it. I see you opened https://github.com/django/django/pull/7849.

**Pull request closed without merging**